### ISSUE TITLE
Query optimization

### ISSUE THREAD

**Issue by borisxm** [sraoss/grafana-sqldb-datasource#3]
I've tested this excellent datasource with PostgreSQL 9.5 and found small data queries to be fast enough. But larger ones (for periods like 6 months and year on 5 minute series), take significant time on the average hardware. After some profiling a way to slightly improve things was found.

Initially the query for six months looks like this:

```
SELECT round(extract(epoch from evtime::timestamptz) / 2160) * 2160 * 1000 AS time_msec, avg(value) 
FROM public.ambient WHERE evtime > to_timestamp(1453827578)
GROUP BY round(extract(epoch from evtime::timestamptz) / 2160) * 2160
ORDER BY round(extract(epoch from evtime::timestamptz) / 2160) * 2160;
```

It returns 6798 rows in 740-740mS. One may note, that expression "round(extract(epoch from evtime::timestamptz) / 2160) * 2160" calculated multiple times. After replacing expressions in GROUP BY/ORDER BY clauses with time_msec the execution time dropped down to 570-575mS which is more than 20% of improvement. Since we're dealing with integer numbers after round(), an extra 1000 multiplier do not affect result of grouping.

**Comment by anzai** [comment]
Thank you for your comment.

I changed the way to specify items in GROUP BY and ORDER BY. Now the executed query is like this.
```
SELECT round(extract(epoch from evtime::timestamptz) / 2160) * 2160 * 1000 AS time_msec, avg(value) 
FROM public.ambient WHERE evtime > to_timestamp(1453827578)
GROUP BY 1 ORDER BY 1;
```
Does this change improve your queries?

**Comment by borisxm** [comment]
Yes, this did improve particular query time from 420ms down to 250ms. Thank you.

I've looked at the actual query supplied to the server and it now looks like this:
```
SELECT round(extract(epoch from time::timestamptz) / 43200) * 43200 * 1000 AS time_msec, avg(battery_vdc)
FROM public.ups2_1m WHERE time > to_timestamp(1441243837)
GROUP BY round(extract(epoch from time::timestamptz) / 43200) * 43200 * 1000
ORDER BY round(extract(epoch from time::timestamptz) / 43200) * 43200 * 1000
```
Probably optimizer automatically recognizes identical expressions and calculates them only once.

**Comment by szocske42** [comment]
I have "cumulative" (monotonically increasing) interface counters in my database, so I must use window functions to calculate rates. I also have a lot of them. As a result the default queries from the query builder are too slow, so I use queries of this style, see below. Readability would benefit greatly from a "$unixInterval", resolving to "EXTRACT(EPOCH FROM '$interval'::interval)"

Not an easy read, but the gist of it is the generate_series for the desired samples, 
then selecting the closest actual timestamps I have the right data for (ifid is my interface id).
I could have divided by the actual duration between the actual datapoints, but that would make it even harder to read, and not that more accurate.

I have an index on ts, (the timestamp field) and on ifid.
I'm testing on a laptop with ~60 million random records (~2 weeks worth, every interface in a datacenter sampled every 10 seconds) It's fairly responsive still. (I have a handful of graphs on each dashboard, and I show a few more values per graph.)

A cache would be nice though :-)

```
SELECT 
EXTRACT(EPOCH FROM t::timestamptz) * 1000 
    AS time_msec, 
  (rx_bytes - lag(rx_bytes) OVER (ORDER BY t) )::float / EXTRACT(EPOCH FROM '$interval'::interval) 
    AS bytes
FROM reading JOIN 
  (SELECT t FROM generate_series($timeFrom, $timeTo, '$interval'::interval ) t ) AS sample 
  ON reading.ts IN 
  (SELECT ts FROM reading WHERE ts > sample.t and ifid = $ifid ORDER BY ts LIMIT 1) 
WHERE ifid = $ifid and $timeFilter 
  and ts BETWEEN sample.t AND sample.t + '$interval'::interval 
ORDER BY 1

```
